cmd/rook/ceph/operator.go

```diff
@@ -91,14 +111,132 @@ func startOperator(cmd *cobra.Command, args []string) error {
 	if err != nil {
 		rook.TerminateFatal(err)
 	}
-
 	rookImage := rook.GetOperatorImage(context.Clientset, containerName)
 	serviceAccountName := rook.GetOperatorServiceAccount(context.Clientset)
 	op := operator.New(context, volumeAttachment, rookImage, serviceAccountName)
+	if isPresent, err := isSecretPresent(context); isPresent && err == nil {
+		err = initWebhook(context)
+		if err != nil {
+			logger.Errorf("an error has occurred while initializing webhook: %+v", err)
+		}
+	}
+	if err != nil {
+		logger.Errorf("error has occurred: %+v", err)
+	}
 	err = op.Run()
 	if err != nil {
 		rook.TerminateFatal(errors.Wrapf(err, "failed to run operator\n"))
 	}
 
 	return nil
 }
+
+func isSecretPresent(context *clusterd.Context) (bool, error) {
+
+	_, err := context.Clientset.CoreV1().Secrets(namespace).Get(secretName, metav1.GetOptions{})
+	if err != nil {
+		// If secret is not found. All good ! Proceed with rook without admission controllers
+		if apierrors.IsNotFound(err) {
+			return false, nil
+		}
+		return false, errors.Errorf("Error happened while retrieving the Secret: %+v", err)
+	}
+	return true, nil
+}
+
+func initWebhook(context *clusterd.Context) error {
+	// At this point volume should be mounted, so proceed with creating the service and validatingwebhookconfig
+	err := createWebhookService(context)
+	if err != nil {
+		return errors.Errorf("failed to create service: %+v", err)
+	}
+	err = createValidatingWebhookConfig(context)
+	if err != nil {
+		return errors.Errorf("failed to create validatingwebhookconfig: %+v", err)
+	}
+	// As http.ListenAndServe is a blocking call. We are starting our server in its own goroutine so that rook functionality is not affected.
+	go webhook.StartServer(context)
+
+	return nil
+}
+
+func createWebhookService(context *clusterd.Context) error {
+	serviceClient := context.Clientset.CoreV1().Services(namespace)
+
+	webhookService := corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      acServiceName,
+			Namespace: namespace,
+		},
+		Spec: corev1.ServiceSpec{
+			Ports: []corev1.ServicePort{
+				{
+					Port: servicePort,
+					TargetPort: intstr.IntOrString{
+						IntVal: serverPort,
+					},
+				},
+			},
+			Selector: map[string]string{
+				"app": "rook-ceph-operator",
+			},
+		},
+	}
+	_, err := serviceClient.Create(&webhookService)
+	if err != nil && !apierrors.IsAlreadyExists(err) {
+		return err
```

Comment:
wrap error

Reply:
context for this error has been provided in initWebhook func